torch/_inductor/select_algorithm.py

```diff
@@ -366,6 +372,46 @@ def __init__(
         self.template_out: Optional[str] = None
         self.ops_handler: Optional[V.WrapperHandler] = None  # type: ignore[name-defined]
 
+        # The generated code is not dependent on the input nodes names, or symbolic sizes names.
+        # However, some of the variables returned by generate_and_load that are computed during the
+        # triton template expansions (code generation) are dependent on those.
+        # In order to cache the code generation and avoid redoing it for similar inputs that varies only by
+        # input names or symbol names, we do a record and replay method.
```

Comment:
as @drisspg pointed out, we also need to be careful about global/input state. for instance, these freezing of layouts:

https://github.com/pytorch/pytorch/blob/6fb089f2a2eea75a45ac2340f0e68736524e20bf/torch/_inductor/select_algorithm.py#L922-L927

Reply:
I am unsure how to capture this, maybe add a counter to the numbers of freezed inputs and add to the state?
but we have to capture all such global input mutations  
what do you guys think?
in addition to that i think it make sense to make this opt in. I will make it opt-in in two levels. 

  1. opt in with respect to the template (make the mm_template  only for now opt into this)
  2. opt in as not enabled by default also.
  
 
 what do you guys think
 @eellison @drisspg

Reply:
addressed as explained,